### DISCUSSION TITLE
Unsupported hardware

### DISCUSSION THREAD

**Discussion by 3hhh** [corrados/edrumulus#81] [Q&A] [answered]
Hi again @corrados,

I was wondering how one adds support for currently unsupported hardware to edrumulus?

If I see it correctly, you usually [measure some input signal from the hardware](https://github.com/corrados/edrumulus/tree/main/algorithm/signals) (how's that done? any hardware required?) and then feed it into your [Octave model](https://github.com/corrados/edrumulus/blob/main/algorithm/edrumulus.m) to enable quick testing for [the required parameters](https://github.com/corrados/edrumulus/blob/main/edrumulus_parameters.cpp)?

Maybe I'll give it a shot, if my [Millenium mod](https://github.com/corrados/edrumulus/discussions/80) works somehow (yet tbd).

Thanks again in advance for your answer!

**Comment by corrados** [comment, marked as answer]
The way I am doing it is:

- attach the pad to a normal sound card and record the strikes with 8 kHz sampling rate and according to https://github.com/corrados/edrumulus/blob/main/doc/algorithm.md#test-signals
- add the recording to https://github.com/corrados/edrumulus/tree/main/algorithm/signals
- I create a new type in https://github.com/corrados/edrumulus/blob/main/algorithm/signalsandsettings.m and use an initial parameter set from a similar pad as a starting point
- I then call the script signalsandsettings.m in Octave which automatically uses the script drumtrigger.m
- I adjust the parameters so that the decay curve looks ok and try to get as little false triggers as possible
- then these parameters can be used in https://github.com/corrados/edrumulus/blob/main/edrumulus_parameters.cpp
- to get it fully working, the new type has to be added at multiple places, e.g. in the GUI python script, too

**Reply by 3hhh** [reply]
> https://github.com/corrados/edrumulus/blob/main/doc/algorithm.md#test-signals

I guess the process is somehwat different for kicks, cymbals and Hihats, but I guess I get the idea, thanks!

**Comment by corrados** [comment]
Your modified Millenium drum pad will be very similar to the PD-120 or PD-80R (PD-85). Therefore, I would start using one of these. Maybe it is not necessary to create a new type. Just play with the trigger settings and see if that already works fine for you.

**Comment by 3hhh** [comment]
@corrados Would you mind explaining the [available parameters](https://github.com/corrados/edrumulus/blob/f63b367f78c7ee1b8e91e68009a4e7668ca356ec/edrumulus_parameters.cpp#L24-L56) a bit more than in the cpp file?
I'd like to add support for my Pads and Cymbals... So I guess I need to learn it. I do have some understanding of DSP, but it's been a while that I needed it...

In particular I played around with the decay curve a bit, noticed that it's always linear in the Octave model (I guess cymbals cannot be handled by the model?) and noticed that only the second decay appears to have any impact in the model. What are the others for then?

Is mask time the time for which any triggering is prevented after a peak is detected?

I also guess that the hot spot attenuation (whatever that is, idk) currently only exists in the model, but not in the cpp code?

Btw this is how my first 5 hits look like as interpreted by the default pad (I just applied a 20 db threshold instead of 17db):
![sample](https://github.com/corrados/edrumulus/assets/13033030/d682cdc3-6948-46d4-96f8-4d684d0e634e)

**Reply by corrados** [reply]
> In particular I played around with the decay curve a bit, noticed that it's always linear in the Octave model (I guess cymbals cannot be handled by the model?) and noticed that only the second decay appears to have any impact in the model. What are the others for then?

The decay curve can have up to three segments. But most pads only use one segment (mostly the middle segment). Each segment defines a length and a gradient. If the length is set to 0, the segment is disabled.

Three segments are only needed in very special cases like the VH12 where we have to suppress ringing of the two halves of the pad in case it is half open. You can see this ringing in the following plot:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/130b15c7-27e8-438e-8b39-7a0e9cc6600f)

> Is mask time the time for which any triggering is prevented after a peak is detected?

Yes, it is.

> I also guess that the hot spot attenuation (whatever that is, idk) currently only exists in the model, but not in the cpp code?

Yes, hot spot detection never worked correctly.

You can create your own pad type but as I mentioned above, I think the PD-120 pad type should work pretty good for your pad, too.

**Comment by 3hhh** [comment]
Thanks once again!

Here are my preliminary results:
![prel](https://github.com/corrados/edrumulus/assets/13033030/3fbdd1a3-36f4-4243-89da-ced9039602e8)

It doesn't look too bad I guess, but two things strike me:
1. The velocity goes high really quick. I had noticed edrumulus to almost only trigger high velocity hits when I tested it without any modifications.
2. At low velocity positional sensing doesn't seem to work too well.

Are there any parameters to fix those two issues?

Also, what's to respect with regards to the press rolls?
Are these only to estimate the mask time?

**Reply by corrados** [reply]
> The velocity goes high really quick. I had noticed edrumulus to almost only trigger high velocity hits when I tested it without any modifications.

That's a weird behavior. I haven't seen this with my pads. You can select velocity curves for each pad. Maybe you can compensate a bit for this behavior by using LOG or EXP.

> At low velocity positional sensing doesn't seem to work too well.

This usually happens if you do not detect the correct first peak. You have to zoom in to see the individual peaks to tweak the parameters.

> Also, what's to respect with regards to the press rolls?

I use these to fine-tune the parameters. Again, you have to zoom in to each peak to see if the detected windows (pre-scan, scan, mask) sit on the correct place.

**Reply by corrados** [reply]
If you zoom in at the beginning of the detected peak, you see something like the following:
![image](https://github.com/corrados/edrumulus/assets/46655886/e78255dd-c025-4d24-83ae-d127462112ce)
The green dot must be on the first peak, otherwise the positional sensing will not work correctly.

**Reply by 3hhh** [reply]
Are you implying that a dark blue dot (the one on the first peak before the main peak) as in my situation will break positional sensing?

Also, wrt rim shots I guess the green curve is the relevant one? Unfortunately I don't see it at all in my samples...
What does the rim shot algorithm assume / how does it work? I'm not sure how Millenium mixes the second Piezo into the signal.

I guess I'll open a draft PR with my samples so that you can have a look at them, too. Maybe you see something super obvious.

EDIT: PR: https://github.com/corrados/edrumulus/pull/89

**Reply by corrados** [reply]
What dark blue dot you are referring to? I can see a black, green and magenta dot in my figure. I am referring to the green dot which is placed on the first main peak.

To see the rim piezo channel, you have to create a stereo wave file where one channel is the mesh head piezo and the other contains the rim sensor (as far as I remember, the rim channel is on the right stereo channel).

Have you played with the rim shot threshold already? Didn't that work for you?

**Reply by 3hhh** [reply]
I was referring to the many dark blue dots on my diagram [above](https://github.com/corrados/edrumulus/discussions/81#discussioncomment-6753392).

> To see the rim piezo channel, you have to create a stereo wave file where one channel is the mesh head piezo and the other contains the rim sensor (as far as I remember, the rim channel is on the right stereo channel).

Ok I didn't record the second channel. That should explain why I'm not seeing any rim shots being detected...

However since I don't have leftover I/O ports for stereo signals anyway, I guess I'll only enable rim shots on the snare. I'll look into the snare signal next then.

**Reply by 3hhh** [reply]
The results for the snare look very promising. In particular the dynamics and pos sensing looks a lot better than with the tom. :-)

Unsurprisingly I didn't have to modify the parameters - it's pretty much the same piece of hardware after all.

I think the difference wrt dynamics & pos sensing comes from the piezo foam which is pretty much the only different part than with the tom. The snare foam is much softer than the one of the tom. Maybe I'll try replacing the foam on the toms and have another look at the results (EDIT: Tried that and the signals unfortunately don't look much better. Maybe the Piezo is simply too close to the mesh head to provide really good results, idk).

Rim shots look pretty bad according to the diagram, but I think you already mentioned in your doc that the algorithm isn't too great.

I haven't tested in reality yet...

**Reply by corrados** [reply]
Please write a feedback how good the rim shots perform when you try it in your real system.

**Reply by 3hhh** [reply]
Hm I tested Tom and Snare against an edrumulus prototype 2 with a Teensy 4.1 on the command-line UI and unfortunately the results are pretty bad atm:
- Both require a velocity threshold of around 20 to not get constant noise signals (velocity 1 and sometimes even higher signals constantly scrolling down the screen).
- The velocity sensitivity then needs to be raised to 15 for the tom to get any reasonable dynamics. Even the maximum of 31 doesn't produce any reasonable dynamics for the snare. Soft hits aren't detected on either.
- Pos sensing is all over the place for the tom. For the snare there seems to be some rate of accuracy, but it's still only 70% or so. In particular hitting the upper edge usually results in a middle detection.
- I haven't tested rim shots.

In total the model doesn't seem to live up to its expectations in practice. :-(

**Reply by corrados** [reply]
> Both require a velocity threshold of around 20 to not get constant noise signals (velocity 1 and sometimes even higher signals constantly scrolling down the screen).

I don't have a Teensy 4.1 available, only a Teensy 4.0. To my impression, the Teensy 4.0 has less issues with noise compared to the ESP32 board. For the ESP32 board, sometimes the USB power supply was responsible for a lot of noise. You could try to use a powered USB hub. Also you could try to use a different USB port on your PC. I hope that the voltage converter on the Teensy 4.1 is not worse than the one on the Teensy 4.0.

What you also can do is to try to mitigate the impulse noise in software by using the a "spike" value of 4 in the Edrumulus command line GUI.

> The velocity sensitivity then needs to be raised to 15 for the tom to get any reasonable dynamics. Even the maximum of 31 doesn't produce any reasonable dynamics for the snare. Soft hits aren't detected on either.

As you can see in the edrumulus_paramters.cpp file, my velocity_sensitivity values are all pretty small. I have never needed such high values. Please check your schematics once more. Maybe the diodes are in reversed direction or a connection on the board is incorrect.

> Pos sensing is all over the place for the tom. For the snare there seems to be some rate of accuracy, but it's still only 70% or so. In particular hitting the upper edge usually results in a middle detection.

What is the hardware difference between your tom and snare pads? Have you modified both of them? If you have only modified the snare pad but the tom pad has its head sensor still close to the rim, positional sensing will not work the same as with center mounted piezo sensors. On the PDX-8 I had some positional sensing working but it was much worse than on center mounted piezos.

**Comment by 3hhh** [comment]
> You could try to use a powered USB hub.

You were right about the powered hub. That just enabled me to lower the threshold to 12 (vs 20 before) and receive at least some soft hits on the snare. However I also had to modify other parameters to rather absurd values (mask time = 0, spike = 2), which leads me to the conclusion that the model is way different from reality.

Is it possible to receive the original signal directly from the Teensy?
I tried to use [this code](https://github.com/janscience/TeeRec/blob/main/examples/audiorecorder/audiorecorder.ino), but without success.

I'm pretty sure my original recording for the model had some software interpolations in between.

> What is the hardware difference between your tom and snare pads? Have you modified both of them?

Yes, I modified both. The main difference is that the piezo position of the snare is lower than that of the tom (by 2-3mm or so). Also, the snare piezo has more and softer trigger cushion foam on top of it than the tom.
I'll try replacing the foam by more professional one...

**Reply by corrados** [reply]
> However I also had to modify other parameters to rather absurd values (mask time = 0, spike = 2), which leads me to the conclusion that the model is way different from reality.

The mask time should not have an effect on the soft hits detection. It should only prevent double-triggers. Some other things seems to go wrong here.

> I'm pretty sure my original recording for the model had some software interpolations in between.

What do you mean be software interpolations?

I was looking at the signals in your pull request: First, **Tom1 signal**:

![grafik](https://github.com/corrados/edrumulus/assets/46655886/8eb040f0-d716-483b-9f91-0eca477ab5a7)

You configured the second gradient steeper than the first one. I usually don't do it. For your signal I would say that a single straight line should be sufficient. Also, the decay starts at a too high value. You could lower it.

Your first peak looks overloaded:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/ed6b76c3-810d-47a3-82b8-2163f5ccee71)

Interestingly, even your low level strikes look strange:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/70a19852-290c-4ade-8c3f-51d28cc5aa35)

**Snare signal**:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/01b6aa10-8e30-4877-b80b-8b3aaefacd59)

You have three almost identical main peaks in the blue trace. That's strange. This is quite unusual. That calls for a longer mask time but it seems from the plot that it is already long enough. The first peak is also not a clean peak but has a lot of smaller peaks or ripple. I am not sure what causes this but I would expect that the positional sensing algorithm will have problems with that.

**Reply by 3hhh** [reply]
>> I'm pretty sure my original recording for the model had some software interpolations in between.

> What do you mean be software interpolations?

It was passed from one pulseaudio server to another where I recorded it. So probably there was at least some downsampling involved as I doubt that the second server forwarded my recording sample rate to the first (my OS probably limits that communication).

In the meantime I managed to get my hands on the raw ADC input signal via [this code](https://gist.github.com/3hhh/f1c4a028b053bb2ceaf380ef7565762d).
Admittedly it also only runs at 44kHz.

Should the `signalsandsettings.m` also work for that lower amplitude signal coming from the ADC?

I still have to look into more detail in your comments from above.

**Reply by corrados** [reply]
> Should the signalsandsettings.m also work for that lower amplitude signal coming from the ADC?

Sure, you can simply change the scaling in [this line](https://github.com/corrados/edrumulus/blob/dc5fa32316835a920a8efb32c7fef063bf283f6c/algorithm/signalsandsettings.m#L75).

**Reply by 3hhh** [reply]
Thanks, I see. Nevertheless I'm rather confused why I need to do this at all.

The signal I measured is roughly 40dB weaker than the signal I measured at my line in.
If I understand the prototype 2 circuit correctly, it is supposed to get the peizo signal in the 3V3 range of the ADC. If it does that correctly, imho there shouldn't be any scaling necessary at all?

So I guess I was either doing something very wrong or my circuit is somewhat bad.

Btw it also looks pretty noisy...

Also, I read that piezos operate in the +-10V range, but the line-in works at +-1V range. ~Why was I not seeing any clipping on the line-in then?~ EDIT: After measuring on the line-in again, I can say that there is clipping for sure. It's just not so obvious if you regulate the input volume down (which appears to happen in software and thus there's clipping anyway). So imho it's impossible to generate a test signal without clipping unless one only uses very soft hits.

**Reply by corrados** [reply]
> If it does that correctly, imho there shouldn't be any scaling necessary at all?

Are you referring to the scaling in the Octave file? In the C++ code which is loaded on the micro controller, there is no scaling.

> Btw it also looks pretty noisy...

Yes, I know. See, e.g., [this](https://github.com/corrados/edrumulus/blob/main/doc/ChangeLog.md#2021-04-04-adc-spikes-of-the-esp32).

**Reply by 3hhh** [reply]
> > If it does that correctly, imho there shouldn't be any scaling necessary at all?
> 
> Are you referring to the scaling in the Octave file? In the C++ code which is loaded on the micro controller, there is no scaling.

I was referring to the signal seen in the recorded wav file. Since it directly comes from the input circuit, it should be in the correct range, shouldn't it?

Anyway I did a small simulation of the edrumulus prototype 2 voltage divider circuit in `simulide` and it showed that the circuit is designed for around 10V piezo input. However for 1V piezo input range, the signal amplitude (and thus the SNR to the ADC) will be really low. This is pretty much what I'm experiencing imho. Apparently replacing the 22 kOhm resistors with higher ones (0,5 MOhm or so) should get me a better input signal. However it all depends on the output strength of the piezo... Is there any simple way to measure that? I don't want to fry the teensy...

**Reply by corrados** [reply]
I am the wrong guy to ask. I'm not good at hardware design. I selected the resistors in a way that I got a good compromise between sensitivity and overload with the pads I own. And yes, I intentionally drive very loud signals in the overload to get better dynamics. I spent a good amount of time in developing a clipping compensation algorithm in software. This works fine for me.

**Reply by 3hhh** [reply]
Thanks for the clarification!

Ok, then I'll have to find a way on my own. I happen to know 1-2 skilled people in electronics...

I'll let you know if the results work or not.

**Reply by 3hhh** [reply]
> The signal I measured is roughly 40dB weaker than the signal I measured at my line in. If I understand the prototype 2 circuit correctly, it is supposed to get the peizo signal in the 3V3 range of the ADC. If it does that correctly, imho there shouldn't be any scaling necessary at all?
> 
> So I guess I was either doing something very wrong or my circuit is somewhat bad.

Omg... I was chasing gosts there... The Teensy ADCs run at 10 bits by default, the wav is saved at 16 bits per sample... So you'll always have 30-40dB (6 bit) seemingly reduced signal amplitude.

**Reply by 3hhh** [reply]
My latest tests with https://github.com/corrados/edrumulus/pull/92 show that I get the best signal with 10M Ohm instead of 22k Ohm resistors.
This would indicate that my Piezo signal is in the range of 0.1 to 1V, which is really unexpected. I'll try different trigger cushion foam...

According to my `simulide` simulation matching resistor values are:
max. input amplitudes --> resistors:
0.1-1V --> 10M Ohm
1-4V   --> 500k Ohm
4-10V  --> 100k Ohm
10-30V --> 22k Ohm

**Reply by 3hhh** [reply]
> I'll try different trigger cushion foam...

That didn't change anything with regards to SNR. I hope it improves pos sensing, but I'll check that later.

Anyway the Millenium toms seem to feature a circuit inside each Tom that lowers the input range. At least that's what my previous results indicate. Also, I found two 10-15k resistors and a poti inside it. Moving the latter gives insignificantly better results.
There's also a lot of electrical cross-talk between the hit (left/tip) and rim (right/rear) channel. One can e.g. disconnect the rim piezo and will still see the hit signal on the right channel - albeit at a somewhat lower amplitude.

I'll test whether I get good results with the 10M resistors and otherwise either remove the Millenium circuit or give up...

**Reply by 3hhh** [reply]
> I'll test whether I get good results with the 10M resistors and otherwise either remove the Millenium circuit or give up...

Results are okayish wrt velocity, bad wrt pos sensing. Removing the Millenium circuit doesn't seem to change the incoming signal.

**Reply by corrados** [reply]
That's interesting. So, it seems that there is more than just the requirement of a center head-mounted piezo to get a good positional sensing performance. [thijstriemstra](https://github.com/thijstriemstra) has a Drum-tec Diabolo which is supposed to have positional sensing support but the positional sensing performance in Edrumulus was never good as you can [read in this thread](https://github.com/corrados/edrumulus/discussions/30). Maybe the type of mesh-head has also an influence on the positional sensing.

Unfortunately, I did not have a Drum-tec Diabolo pad myself but I asked in the official Drum-tec Facebook group and they confirmed that positional sensing works fine with the Roland modules. There are just a few Youtube videos about this pad and they even show a bit of positional sensing but it is hard to judge if the positional sensing performance is really as good as with original Roland mesh-head pads. Judging from the Youtube videos, I think some positional sensing works but the performance is not really good. It seems for the Diabolo pad and a Roland module, the positional sensing does not have a nice transition from edge to middle. It seems to jump from one state to the other.

Would be interesting to see if eDRUMin performs better with these pad with regard to positional sensing.

**Reply by 3hhh** [reply]
Well, the pos sensing in the model looks relatively fine. In reality however I tend to get relatively reliable middle sensing and bad edge sensing (at the right edge, it jumps to 100% middle or 100% right edge, somewhere in between it's usually 100% middle; at the top edge it's pretty reliably always 100% middle).

I'd like to use the `signalsandsettings.m` model on a [real signal](https://github.com/corrados/edrumulus/pull/92), but currently I don't understand the model well enough to adapt it to real signals.

**Reply by 3hhh** [reply]
New findings wrt resistor values: https://github.com/corrados/edrumulus/issues/93

Also, I noticed that soldering female 6,3mm jack connectors are apparently a bad choice for edrumulus. Using male-male 6,3mm cables, cutting off one end and soldering male-male jumper cables to the ends appears to be better (less noise, no issues with touching the cables etc.).

**Reply by AlienDiva** [reply]
Good to know, I've ordered a roll of wide.copper tape for my build just in case. Everything is an antenna.

**Comment by 3hhh** [comment]
So I finally made it to the cymbals of my MPS 750x set.

Unfortunately it looks like I hit into another road block:
The ride is a 3-zone cymbal, the crash a 2-zone, both with only a single stereo 6,3inch output. A 5-hit bow, bell, edge line-in sample of the ride looks like this:
![ride_regular_bell_edge](https://github.com/corrados/edrumulus/assets/13033030/d24bb812-248d-4d90-af41-cae6903333aa)

So it has three modes, but looks like mono? I guessed it is a rim switch and configured it that way, but unfortunately edrumulus doesn't recognize the edge. Do you happen to have any ideas @corrados ?

To me it looks like there's some low amplitude signal encoded in at least one or maybe both channels.

**Reply by corrados** [reply]
Maybe it is a hardware design similar to Yamaha cymbals: https://github.com/RyoKosaka/HelloDrum-arduino-Library/blob/master/docs/circuit.md

**Reply by 3hhh** [reply]
Hm I thought the lower channel might just be GND on R, tried it as such and measured the other channel (S) and found a signal that looks very much like the upper signal (T) - unfortunately for all types of hits. I'm a bit at a loss about that design...

Anyway thanks for the link, maybe I'll find something related.

**Reply by 3hhh** [reply]
The [Yamaha circuit](https://camo.githubusercontent.com/e9c1db572db526b9fe68adc4ec171a6e06c3ce9ce3a7e62a031599c536cce34f/68747470733a2f2f6f70656e2d652d6472756d732e636f6d2f696d616765732f636972637569742f3037342f63796d62616c2e706e67) connects 5V to the ring, which indicates that the author believes the Roland & Yamaha cymbals to be powered. This is very different from the edrumulus circuit. Do you power your Roland cymbals / what circuit do you use?

**Reply by corrados** [reply]
I use the same schematic for the rim switch as for the piezo. Since I use a DC coupling (no serial capacitor), the rim switch does short cut the input and this modifies the measured voltage just a tiny bit. But this tiny bit is sufficient for good detection results.

**Reply by 3hhh** [reply]
I see.

Anyway the MSP 750x cymbals look like they don't use switches.

By now I believe it's a simple differential signal with `rim = |T - R|` and `bow = |S - T|`. Probably `bell = |S - R|`, but I haven't verified that yet. The drum computer probably just checks which of the three signals rings and plays the respective note.

I think edrumulus doesn't support that, does it? It just checks whether the rim piezo reaches a threshold and plays the rim sound then, doesn't it?

**Reply by corrados** [reply]
Yes, exactly.

**Reply by 3hhh** [reply]
After further investigation it seems that the bell works differently (maybe a switch, idk).

I guess using the prototype 2 circuit with T as GND should get at least the rim going as the bow signal also works for bell & rim (it always plays).

**Comment by 3hhh** [comment]
@corrados
How can I measure the hihat controller?
Its resistance doesn't change on hihat movements according to my multimeter. I noticed that the Millenium drum module seems to power the stereo jack input with 5V and 2,5V.
So it's some sort of active sensor?

Btw crash & ride are powered with 3V3 as well.

**Reply by corrados** [reply]
On Roland hihat controllers, it's simply a potentiometer. I have no idea what it is in your Millenium controller.
